src/kvdb/prometheus.rs

```diff
@@ -0,0 +1,135 @@
+// Copyright 2022 Parity Technologies (UK) Ltd.
+// This file is part of polkadot-introspector.
+//
+// polkadot-introspector is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// polkadot-introspector is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with polkadot-introspector.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::kvdb::IntrospectorKvdb;
+use clap::Parser;
+use color_eyre::Result;
+use log::info;
+use prometheus_endpoint::{prometheus::IntGaugeVec, Opts, Registry};
+use std::sync::Arc;
+use tokio::sync::Mutex;
+
+#[derive(Clone, Debug, Parser, Default)]
+#[clap(rename_all = "kebab-case")]
+pub struct KvdbPrometheusOptions {
+	/// Prometheus endpoint port.
+	#[clap(long, default_value = "65432")]
+	port: u16,
+	/// Database poll timeout (default, once per 5 minutes).
+	#[clap(long, default_value = "300.0")]
+	poll_timeout: f32,
+}
+
+struct KvdbPrometheusMetrics {
+	keys_size_gauge: IntGaugeVec,
+	values_size_gauge: IntGaugeVec,
+	elements_count_gauge: IntGaugeVec,
+}
+
+pub async fn run_prometheus_endpoint_with_db<D: IntrospectorKvdb + Send + Sync + 'static>(
+	db: D,
+	prometheus_opts: KvdbPrometheusOptions,
+) -> Result<Vec<tokio::task::JoinHandle<()>>> {
+	let prometheus_registry = Registry::new_custom(Some("introspector".into()), None)?;
+	let metrics = Arc::new(Mutex::new(register_metrics(&prometheus_registry)));
```

Comment:
Do we really need the mutex here?

Reply:
No, we even don't need an arc, as gauges use atomic variables internally.